src/snowflake/snowpark/modin/plugin/_internal/binary_op_utils.py

```diff
@@ -272,14 +291,74 @@ def compute_binary_op_between_snowpark_columns(
             binary_op_result_column = pandas_lit(False)
         else:
             binary_op_result_column = first_operand.equal_null(second_operand)
+    elif (
+        op in ("rsub", "sub")
+        and isinstance(first_datatype(), TimestampType)
+        and isinstance(second_datatype(), NullType)
+    ):
+        # Timestamp - NULL or NULL - Timestamp raises SQL compilation error,
+        # but it's valid in pandas and returns NULL.
+        snowpark_pandas_result_type = NullType()
+        binary_op_result_column = pandas_lit(None)
+    elif (
+        op in ("rsub", "sub")
+        and isinstance(first_datatype(), NullType)
+        and isinstance(second_datatype(), TimestampType)
+    ):
+        # Timestamp - NULL or NULL - Timestamp raises SQL compilation error,
+        # but it's valid in pandas and returns NULL.
+        snowpark_pandas_result_type = NullType()
+        binary_op_result_column = pandas_lit(None)
+    elif (
+        op == "sub"
+        and isinstance(first_datatype(), TimestampType)
+        and isinstance(second_datatype(), TimestampType)
+    ):
+        if (
+            first_datatype().tz is TimestampTimeZone.NTZ
+            and second_datatype().tz is TimestampTimeZone.TZ
+        ) or (
+            first_datatype().tz is TimestampTimeZone.TZ
+            and second_datatype().tz is TimestampTimeZone.NTZ
+        ):
+            raise TypeError(
+                "Cannot subtract tz-naive and tz-aware datetime-like objects."
+            )
+        snowpark_pandas_result_type = TimedeltaType()
+        binary_op_result_column = iff(
+            is_null(first_operand).__or__(is_null(second_operand)),
+            pandas_lit(native_pd.NaT),
+            datediff("ns", second_operand, first_operand),
+        )
```

Comment:
I think you can combine `rsub` and `sub` as follows (sorry couldn't find a way to delete `rsub` case b/c of the other comment splitting the code):

```suggestion
    elif (
        op in ("rsub", "sub")
        and isinstance(first_datatype(), TimestampType)
        and isinstance(second_datatype(), TimestampType)
    ):
        if (
            first_datatype().tz in [TimestampTimeZone.TZ, TimestampTimeZone.NTZ]
            or second_datatype().tz in [TimestampTimeZone.TZ, TimestampTimeZone.NTZ]
        ):
            raise TypeError(
                "Cannot subtract tz-naive and tz-aware datetime-like objects."
            )
        snowpark_pandas_result_type = TimedeltaType()
        binary_op_result_column = iff(
            is_null(first_operand).__or__(is_null(second_operand)),
            pandas_lit(native_pd.NaT),
            datediff("ns", second_operand, first_operand),
        )
```